tests/test-buttercup.el

```diff
@@ -613,264 +613,265 @@
-                 :to-have-been-called-with
-                 (make-list-of-closures '(test-function 1 2 3)))
-                :to-equal
-                (cons nil
-                      "Expected `test-function' to have been called with (1 2 3), but it was not called at all")))
-
-      (it "returns false if the spy was called with different arguments"
-        (test-function 3 2 1)
-        (expect (buttercup--apply-matcher
-                 :to-have-been-called-with
-                 (make-list-of-closures '(test-function 1 2 3)))
-                :to-equal
-                (cons nil
-                      "Expected `test-function' to have been called with (1 2 3), but it was called with (3 2 1)")))
-
-      (it "returns true if the spy was called with those arguments"
-        (test-function 1 2 3)
-        (expect (buttercup--apply-matcher
-                 :to-have-been-called-with
-                 (make-list-of-closures '(test-function 1 2 3)))
-                :to-be
-                t)))
-
-    (describe ":to-have-been-called-times matcher"
-      (before-each
-        (spy-on 'test-function))
-
-      (it "returns error if the spy was called less than expected"
-        (expect (buttercup--apply-matcher
-                 :to-have-been-called-times
-                 (make-list-of-closures '(test-function 1)))
-                :to-equal
-                (cons nil
-                      "Expected `test-function' to have been called 1 time, but it was called 0 times")))
-
-      (it "returns error if the spy was called more than expected"
-        (test-function)
-        (test-function)
-        (expect (buttercup--apply-matcher
-                 :to-have-been-called-times
-                 (make-list-of-closures '(test-function 1)))
-                :to-equal
-                (cons nil
-                      "Expected `test-function' to have been called 1 time, but it was called 2 times")))
-
-      (it "returns true if the spy was called the expected number of times"
-        (test-function)
-        (test-function)
-        (expect (buttercup--apply-matcher
-                 :to-have-been-called-times
-                 (make-list-of-closures '(test-function 2)))
-                :to-equal t))
-
-      (it "use plural words in error message"
-        (test-function)
-        (test-function)
-        (expect (buttercup--apply-matcher
-                 :to-have-been-called-times
-                 (make-list-of-closures '(test-function 3)))
-                :to-equal
-                (cons nil
-                      "Expected `test-function' to have been called 3 times, but it was called 2 times")))
+  ;; We use `before-all' here because some tests need to access the
+  ;; same function as previous tests in order to work, so overriding
+  ;; the function before each test would invalidate those tests.
+  ;; Unfortunately there's no way to do this lexically, so these
+  ;; function definitions are leaked after the tests run.
```

Comment:
Is it possible to lexically bind a symbol's function slot? I don't actually know.

Reply:
I tested this in *scratch*: 
```
(cl-letf (((symbol-function 'my-test-function) nil))
  (fset 'my-test-function (lambda () "1"))
  (my-test-function))
"1"
(my-test-function) => (void-function my-test-function)
```
the macroexpansion is actually rather straightforward.

Reply:
Ok, I'll add that in, then.

Reply:
Quick test in real code shows it wasn't that simple :(
It's probably as you say, it's not possible to lexically bind the function slot.

Reply:
Actually, looking at the macro expansion, isn't that dynamic binding? It sets the symbol function at the start of the block and then reverts it to its old value when the block exits. Since the `describe` block only defines the test and doesn't run it, the dynamic binding will be long gone when the tests actually run, having no effect.

Reply:
Wasn't someone working on an `around-all` macro or something? I think that would probably be what's needed here.

Reply:
I was hoping to not have to reindent that big chunk of code :)
But never mind, we can go ahead with this part of the change as well.

Reply:
I pushed an alternative to #147, what do you think?  It's not pretty, but it does not "leak" and test functions.
Should be possible to convert to some sort of `:func`argument to `describe`.

Reply:
Yes, I think that should work. Perhaps instead of `:func` you could have a `:letf` argument to `describe` that wraps the test-running code in the equivalent `cl-letf` form? Alternatively, I think the most general solution would be to implement `around-each` and `around-all` macros. I might take a stab at that some time soon. For now, though, I think your solution is enough.

Reply:
I just realized `cl-flet` and `cl-labels` are supposed to be lexically scoped. (They actually work by rewriting the function names to different, unused, symbols.) I'll test things out using those and see if it works.

Reply:
Never mind, these actually work by lexically binding symbols normally using `let` and then rewriting calls to the bound functions to use `funcall` to access the symbol value rather than the symbol function. So that won't work.